File: Palace/MyBooks/TPPMyBooksDownloadInfo.m
#import "TPPMyBooksDownloadInfo.h"

@interface TPPMyBooksDownloadInfo ()

@property (nonatomic) CGFloat downloadProgress;
@property (nonatomic) NSURLSessionDownloadTask *downloadTask;
@property (nonatomic) TPPMyBooksDownloadRightsManagement rightsManagement;
@property (nonatomic) TPPMyBooksSimplifiedBearerToken *bearerToken;

@end

@implementation TPPMyBooksDownloadInfo

- (instancetype)initWithDownloadProgress:(CGFloat const)downloadProgress
                            downloadTask:(NSURLSessionDownloadTask *const)downloadTask
                        rightsManagement:(TPPMyBooksDownloadRightsManagement const)rightsManagement {

  return [[[self class] alloc]
          initWithDownloadProgress:downloadProgress
          downloadTask:downloadTask
          rightsManagement:rightsManagement
          bearerToken:self.bearerToken];
}

- (instancetype)initWithDownloadProgress:(CGFloat const)downloadProgress
                            downloadTask:(NSURLSessionDownloadTask *const)downloadTask
                        rightsManagement:(TPPMyBooksDownloadRightsManagement const)rightsManagement
                             bearerToken:(TPPMyBooksSimplifiedBearerToken *)bearerToken
{
  self = [super init];
  if(!self) return nil;
  
  self.downloadProgress = downloadProgress;
  
  if(!downloadTask) @throw NSInvalidArgumentException;
  self.downloadTask = downloadTask;

  self.rightsManagement = rightsManagement;
  self.bearerToken = bearerToken;
  
  return self;
}

- (instancetype)withDownloadProgress:(CGFloat const)downloadProgress
{
  return [[[self class] alloc]
          initWithDownloadProgress:downloadProgress
          downloadTask:self.downloadTask
          rightsManagement:self.rightsManagement];
}

- (instancetype)withRightsManagement:(TPPMyBooksDownloadRightsManagement const)rightsManagement
{
  return [[[self class] alloc]
          initWithDownloadProgress:self.downloadProgress
          downloadTask:self.downloadTask
          rightsManagement:rightsManagement];
}

- (NSString *)rightsManagementString
{
  switch (self.rightsManagement) {
    case TPPMyBooksDownloadRightsManagementUnknown:
      return @"Unknown";
    case TPPMyBooksDownloadRightsManagementNone:
      return @"None";
    case TPPMyBooksDownloadRightsManagementAdobe:
      return @"Adobe";
    case TPPMyBooksDownloadRightsManagementSimplifiedBearerTokenJSON:
      return @"SimplifiedBearerTokenJSON";
    case TPPMyBooksDownloadRightsManagementOverdriveManifestJSON:
      return @"OverdriveManifestJSON";
    default:
      return [NSString stringWithFormat:@"Unexpected value: %ld",
              (long)self.rightsManagement];
  }
}

@end

File: Palace/MyBooks/TPPMyBooksDownloadInfo.swift
import Foundation

@objc enum TPPMyBooksDownloadRightsManagement: Int {
  case unknown
  case none
  case adobe
  case simplifiedBearerTokenJSON
  case overdriveManifestJSON
  case lcp
}

@objc class TPPMyBooksDownloadInfo: NSObject {

  @objc private(set) var downloadProgress: CGFloat
  @objc private(set) var downloadTask: URLSessionDownloadTask
  @objc private(set) var rightsManagement: TPPMyBooksDownloadRightsManagement
  @objc private(set) var bearerToken: TPPMyBooksSimplifiedBearerToken?

  @objc init(downloadProgress: CGFloat, downloadTask: URLSessionDownloadTask, rightsManagement: TPPMyBooksDownloadRightsManagement) {
    self.downloadProgress = downloadProgress
    self.downloadTask = downloadTask
    self.rightsManagement = rightsManagement
    self.bearerToken = nil
    super.init()
  }

  @objc init(downloadProgress: CGFloat, downloadTask: URLSessionDownloadTask, rightsManagement: TPPMyBooksDownloadRightsManagement, bearerToken: TPPMyBooksSimplifiedBearerToken?) {
    self.downloadProgress = downloadProgress
    self.downloadTask = downloadTask
    self.rightsManagement = rightsManagement
    self.bearerToken = bearerToken
    super.init()
  }

  @objc func withDownloadProgress(_ downloadProgress: CGFloat) -> TPPMyBooksDownloadInfo {
    return TPPMyBooksDownloadInfo(
      downloadProgress: downloadProgress,
      downloadTask: self.downloadTask,
      rightsManagement: self.rightsManagement,
      bearerToken: self.bearerToken
    )
  }

  @objc func withRightsManagement(_ rightsManagement: TPPMyBooksDownloadRightsManagement) -> TPPMyBooksDownloadInfo {
    return TPPMyBooksDownloadInfo(
      downloadProgress: self.downloadProgress,
      downloadTask: self.downloadTask,
      rightsManagement: rightsManagement,
      bearerToken: self.bearerToken
    )
  }

  @objc var rightsManagementString: String {
    switch rightsManagement {
    case .unknown: return "Unknown"
    case .none: return "None"
    case .adobe: return "Adobe"
    case .simplifiedBearerTokenJSON: return "SimplifiedBearerTokenJSON"
    case .overdriveManifestJSON: return "OverdriveManifestJSON"
    case .lcp: return "LCP"
    }
  }
}
